### PR TITLE
Sds 131 add list available validators

### PR DESCRIPTION
**Pull request by alanmaydwell**
## Description of change

New "get" endpoint `available_validators` added. This returns details of the validators SDS makes available. Idea is to provide information useful when creating config files.

Three items are included for each validator:

1. `name` - just taken from validator's `__name__` attribute
2. `description` - new function `get_validator_validate_docstring` retrieves the first non-empty line and whole text from each validator's docstring. The first first non-empty line is used as description text.
3. `validator_kwargs` -  the arguments associated with the `valdiator.validate` method. These are extracted using "new" function `get_kwargs_for_filevalidator`, which is based on existing function of same name in `configbuilder.py` tool but has a number of changes.

Response looks like this:

```
[
  {
    "name": "AllowedFileExtensions",
    "description": "Validates that the file extension is in the list of allowed extensions.",
    "validator_kwargs": {
      "extensions": []
    }
  },
  {
    "name": "AllowedMimetypes",
    "description": "Validates that the file mimetype is in the list of allowed mimetypes.",
    "validator_kwargs": {
      "content_types": []
    }
  },
  {
    "name": "DisallowedFileExtensions",
    "description": "Validates that the file extension is not in the list of disallowed extensions.",
    "validator_kwargs": {
      "extensions": []
    }
  },
  {
    "name": "DisallowedMimetypes",
    "description": "Validates that the file mimetype is not in the list of disallowed mimetypes.",
    "validator_kwargs": {
      "content_types": []
    }
  },
  {
    "name": "MaxFileSize",
    "description": "Validates that the file is at most a certain size.",
    "validator_kwargs": {
      "size": 1
    }
  },
  {
    "name": "MinFileSize",
    "description": "Validates that the file is at least a certain size.",
    "validator_kwargs": {
      "size": 1
    }
  }
]
```
###  Limitation
The way the `get_kwargs_for_filevalidator` function retrieves the `validator_kwargs` relies on the validator's arguments being positional, with those with default values on the right (which is usual). This is fine for the current validators and usual for a lot of methods in general. 

However there is possibility that a method could be defined like this: `def validate(a, b, *c, d=1)`.  These type of arguments would not be processed correctly by `get_kwargs_for_filevalidator` which would miss out the `d=1` on the end. This could likely be fixed with extra introspection to pick up  keyword-only arguments but seems not worth the effort as this is a bit of an edge-case that probably won't occur in an actual validator.

### Extra unrelated change
A small bit of housekeeping, highlighted by Adrian. Removed the `configs` directory, which is no longer needed as configs are now stored in a separate repo.

## Link to Jira Ticket

- [SDS-131](https://dsdmoj.atlassian.net/browse/SDS-131)

## Screenshots or test evidence if applicable

At one point had some failures in Postman pipeline run. I think I must have added my new Postman tests to an old version of the collection. Fixed by making sure I had a fresh copy of the Postman collection `json` file and then manually pasted my new tests in. Imported this new file into Postman. Found this one works both locally and passed in pipeline.

image

[SDS-131]: https://dsdmoj.atlassian.net/browse/SDS-131?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/main.py
Comment (adrian-mojd) [inline src/main.py:1]: This naming is much better than have `_router` on the end :) Could you rename the others to match?

File: tests/validation_tests/test_client_configured_validator.py
[the remark this answers is not in the document]
Comment (alanmaydwell) [reply tests/validation_tests/test_client_configured_validator.py:1]: I don't think this is a problem. The `self` parameter is included in one test where it particularly matters but I think it's fine to exclude it from others.

The test  `test_get_kwargs_for_filevalidator_excludes_self_and_file_object_args` has a `validate` method with  both `self` and `file_object` parameters. Its purpose is to verify that these two parameters are excluded from the returned kwargs, so they clearly have to be present for this test.

However, I think it's fine that the other classes I defined for test purposes  have `validate`  methods that lack these arguments. These tests concern the extraction of information from a validator's `validate` method in whatever form it happens to be, not to enforce or verify the presence of particular parameters. I think adding `self` (and possibly `file_object`) to the other tests would just add a bit of clutter.

Is there a problem from doing the above?

[the remark this answers is not in the document]
Comment (adrian-mojd) [reply tests/validation_tests/test_client_configured_validator.py:1]: It passes the tests because the parameters are not used, but it would not work in production code as the first argument is always a reference to the instance, thus giving an unexpected value for `a`. So I think this makes it an unexpected and potentially confusing pattern, and we should add the `self` parameter even though it is not used here.
PyCharm just flags it as an error, but the Python docs put it well: "Note, however, that by not following the convention your code may be less readable to other Python programmers, and it is also conceivable that a class browser program might be written that relies upon such a convention"

[the remark this answers is not in the document]
Comment (alanmaydwell) [reply tests/validation_tests/test_client_configured_validator.py:1]: OK, I'll add `self`, (or maybe `@staticmethod` decorator instead, at least to still cover response from   `get_kwargs_for_filevalidator` when `validate` method has no arguments at all)

File: tests/validation_tests/test_client_configured_validator.py [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+    class HasShortDoc:
+        def validate():
+            "This is the only line"
+            pass
+
+    headline, all = get_validator_validate_docstring(HasShortDoc)
+    assert headline == all == "This is the only line"
+
+
+def test_get_validator_validate_doctring_empty_docstring():
```
Comment (papamaemae) [inline tests/validation_tests/test_client_configured_validator.py:153]: lil typo to fix: change `doctring` to `docstring` - there are a few instances in this file.

File: tests/validation_tests/test_client_configured_validator.py [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+
+
+def test_get_validator_validate_doctring_with_actual_validator():
+    """
+    Just checking ability to digest a genuine validator.
+    Not examining text content to avoid fragility if docstring updated
+    (which is not part of code).
+    """
+    headline, all = get_validator_validate_docstring(MaxFileSize)
+    assert len(headline) <= len(all)
```
Comment (papamaemae) [inline tests/validation_tests/test_client_configured_validator.py:181]: Good range of edge cases covered in these tests and possible docstring examples considered (multi-line, single-line, empty, missing).
I love parametrized tests and find them much easier to read :)
Looks good!

File: Postman/SecureDocStoreAPI.postman_collection.json [hunk 3]
```diff
@@ -206,6 +206,67 @@
...
+						"url": {
+							"raw": "{{SDSBaseUrl}}/available_validators",
+							"host": [
+								"{{SDSBaseUrl}}"
+							],
+							"path": [
+								"available_validators"
+							]
+						},
+						"description": "Expect a 200 OK response from /status"
```
[the remark this answers is not in the document]
Comment (alanmaydwell) [reply Postman/SecureDocStoreAPI.postman_collection.json]: Well spotted. Will Fix.

File: src/main.py
Comment (alanmaydwell) [reply src/main.py:1]: This crossed my mind in the past. Will do.

File: tests/validation_tests/test_client_configured_validator.py
[the remark this answers is not in the document]
Comment (alanmaydwell) [reply tests/validation_tests/test_client_configured_validator.py:1]: Updated

File: src/models/file_validator_spec.py [hunk 4]
```diff
@@ -4,6 +4,8 @@
 class FileValidatorSpec(BaseModel):
     """
     Requires the named file validator with the specified keyword arguments to pass.
+    Although description attribute defaults to None, so does not require specific assingment.
```
[the remark this answers is not in the document]
Comment (alanmaydwell) [reply src/models/file_validator_spec.py:7]: Agree that's clearer. Changed.

File: tests/validation_tests/test_client_configured_validator.py (hunk 2 above)
Comment (alanmaydwell) [reply tests/validation_tests/test_client_configured_validator.py:181]: Thanks. I agree parametrization can make things easier to read.

File: tests/validation_tests/test_client_configured_validator.py
[the remark this answers is not in the document]
Comment (adrian-mojd) [reply tests/validation_tests/test_client_configured_validator.py:1]: Ace, thanks :)
